GetRCConfiguration/handler.ts

```diff
@@ -60,9 +66,84 @@ export const handleEmptyErrorResponse = (configurationId: Ulid) => (
     )
   );
 
+const parseRedisValue: (redisValue: string) => E.Either<Error, Json> = flow(
+  parse,
+  E.mapLeft(() => new Error("Cannot parse RCConfiguration Json from Redis"))
+);
+
+const validateRedisValue: (
+  redisValue: string
+) => E.Either<Error, RetrievedRCConfiguration> = flow(
+  RetrievedRCConfiguration.decode,
+  E.mapLeft(() => new Error("Cannot decode RCConfiguration Json from Redis"))
+);
+
+const parseAndValidateRedisValue: (
+  redisValue: string
+) => E.Either<Error, RetrievedRCConfiguration> = flow(
+  parseRedisValue,
+  E.chain(validateRedisValue)
+);
+
+const handleRedisEmptyResponse: (
+  maybeRedisResponse: O.Option<string>
+) => TE.TaskEither<Error, string> = TE.fromOption(
+  () => new Error("Cannot Get RCConfiguration from Redis")
+);
+
+/**
+ * This method is used to get a configuration, if it exists, by configuratin id.
+ * If the configuration is cached it retrieves it from cache, or else it retrieves
+ * it from cosmosdb and provides to cache it.
+ *
+ * @param configurationId
+ * @returns
+ */
+const getOrCacheMaybeRCConfigurationById = (
+  configurationId: Ulid,
+  redisClient: RedisClientFactory,
+  rccModel: RCConfigurationModel,
+  config: IConfig
+): TE.TaskEither<Error, O.Option<RetrievedRCConfiguration>> =>
+  pipe(
+    getTask(redisClient, `${RC_CONFIGURATION_REDIS_PREFIX}-${configurationId}`),
+    TE.chain(handleRedisEmptyResponse),
+    TE.chainEitherK(parseAndValidateRedisValue),
+    TE.fold(
+      () =>
+        pipe(
+          rccModel.findLastVersionByModelId([configurationId]),
+          TE.mapLeft(
+            e => new Error(`${e.kind}, RCConfiguration Id=${configurationId}`)
+          ),
+          TE.chain(maybeRCConfiguration =>
+            pipe(
+              maybeRCConfiguration,
+              O.fold(
+                () => TE.right(maybeRCConfiguration),
+                rCConfiguration =>
+                  pipe(
+                    setWithExpirationTask(
+                      redisClient,
+                      `${RC_CONFIGURATION_REDIS_PREFIX}-${configurationId}`,
+                      JSON.stringify(rCConfiguration),
+                      config.RC_CONFIGURATION_CACHE_TTL
+                    ),
+                    TE.map(() => maybeRCConfiguration),
+                    TE.orElse(() => TE.of(maybeRCConfiguration))
+                  )
+              )
+            )
+          )
+        ),
+      rCConfiguration => TE.right(O.some(rCConfiguration))
```

Comment:
I think this pipeline is too complex. It can be simplified.

First of all, if you don't want to keep the result of the computation at `119` you can use `TE.chainFirst` instead of `TE.chain`. In this way you don't have to remap the result to the original value.

Once used `chainFirst` maybe you can find an alternative way to `O.fold` to call `setWithExpirationTask`

Reply:
Nice, take a look [b22e966](https://github.com/pagopa/io-functions-service-messages/pull/57/commits/b22e966695d6173bfe9a1c3bde6f9703c88ea9f8)